h/util/uri.py

```diff
@@ -162,6 +164,12 @@ def decode_result(result):
     # Try to extract the scheme
     uri = urlparse.urlsplit(uristr)
 
+    # If the user didn't provide the scheme add it.
+    # This doesn't work on host:port urls.
+    if default_scheme and not uri.scheme and not uristr.startswith("http"):
```

Comment:
> because of cases where the url is encoded like http%3A%2F%2Ffoo%2Ecom since urlparse doesn't handle the encoding nicely.

What concrete cases are you thinking of? The URL should always have been decoded by the time it gets passed to this function.

Reply:
^ There are tests that pass an encoded url in directly so either the tests are wrong or that assumption is not always true. See lines 37 & 38 in tests/h/util/uri_tests.py.
```
    # Should treat messed up urlencoded strings as opaque and return them as
    # is. This is not valid urlencoding and trying to deal with it is going to
    # cause more problems than solutions (for example: is "%2F" in the path
    # section a path segment delimiter or a literal solidus?).
    ("http%3A%2F%2Fexample.com", "http%3A%2F%2Fexample.com"),
    ("http%3A%2F%2Fexample.com%2Ffoo%2F", "http%3A%2F%2Fexample.com%2Ffoo%2F"),
```

Reply:
Interesting. I wonder if the Git history says anything about the origins of this?

Reply:
I'll see if I can find something...

Reply:
I only see one commit from Nick and I find it a little confusing because it seems to be describing behavior that the tests don't in that the tests do not expect normalize to normalize a url encoded url to be the same as a non-encoded url but this commit message seems to indicate that that's the case? Maybe I'm not reading this correctly? 
```
Author: Nick Stenning <nick@whiteink.com>
Date:   Wed Aug 5 15:28:23 2015 +0200

    URI normalisation: normalise percent-encoding
    
    Normalise percent-encoding of paths and query strings as dictated by
    RFC3986.
    
    Paths and query strings can both contain [percent-encoded][1] character
    data. Percent-encoding is a way to include  in URLs (and URIs more
    generally) certain characters which may have semantic meaning in a URL
    (such as "/" in a URL path), or which may simply not be permitted (such
    as literal UTF-8 characters).
    
    Sometimes these percent-encodings must be preserved -- percent-encoded
    UTF-8 unicode, for example. But sometimes URLs that differ only in
    percent-encoding can (according to RFC3986) be considered the same. To
    pick one example, "-" (hyphen) is an "unreserved character", so:
    
        http://example.com/foo-bar, AND
        http://example.com/foo%2Dbar
    
    can be considered the same. In addition, case in percent encodings does
    not carry information, so
    
        http://example.com/foo%2dbar, AND
        http://example.com/foo%2Dbar
    
    are also the same URL.
    
    This commit adds support for normalising percent-encoding of URL paths
    and query strings.
    
    [1]: https://en.wikipedia.org/wiki/Percent-encoding

```